backend/cfg/cfgize.ml

```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
+          | Load _ | Store _ | Intop _ | Intop_imm _ | Negf | Absf | Addf | Subf
+          | Mulf | Divf | Compf _ | Floatofint | Intoffloat | Probe _
+          | Probe_is_enabled _ | Opaque | Begin_region | End_region | Specific _
+          | Name_for_debugger _ )
+      | Call _ | Reloadretaddr | Prologue ->
+        stack_offset, traps, instr
+    in
 
-  let rec update_block : Cfg.t -> Label.t -> handler_stack -> unit =
-   fun cfg label stack ->
     let block = Cfg.get_block_exn cfg label in
     let was_invalid =
-      if block.trap_depth = invalid_trap_depth
+      if block.stack_offset = invalid_stack_offset
       then true
       else begin
-        assert (block.trap_depth = succ (List.length stack));
+        assert (
+          block.stack_offset = compute_stack_offset ~stack_offset ~traps
+          && block.stack_offset >= Proc.trap_size_in_bytes * List.length traps);
         false
       end
     in
     if was_invalid
     then begin
-      block.trap_depth <- succ (List.length stack);
-      let stack, exceptional_successor, body =
-        ListLabels.fold_left block.body ~init:(stack, None, [])
-          ~f:(fun (stack, exceptional_successor, body) instr ->
-            let (stack, exceptional_successor), instr =
-              process_basic exceptional_successor stack instr
+      block.stack_offset <- compute_stack_offset ~stack_offset ~traps;
+      let stack_offset, traps, body =
+        ListLabels.fold_left block.body ~init:(stack_offset, traps, [])
+          ~f:(fun (stack_offset, traps, body) instr ->
+            let stack_offset, traps, instr =
+              process_basic ~stack_offset ~traps instr
             in
-            stack, exceptional_successor, instr :: body)
+            stack_offset, traps, instr :: body)
       in
       block.body <- List.rev body;
-      let (stack, exceptional_successor), terminator =
-        process_terminator exceptional_successor stack block.terminator
+      let stack_offset, traps, terminator =
+        process_terminator ~stack_offset ~traps block.terminator
       in
       block.terminator <- terminator;
       (* non-exceptional successors *)
       Label.Set.iter
-        (fun successor_label -> update_block cfg successor_label stack)
+        (update_block cfg ~stack_offset ~traps)
         (Cfg.successor_labels ~normal:true ~exn:false block);
       (* exceptional successor *)
-      Option.iter
-        (fun (handler_label, handler_stack) ->
+      if block.can_raise
+      then begin
+        assert (Option.is_none block.exn);
+        match traps with
+        | [] -> ()
+        | handler_label :: rest ->
           block.exn <- Some handler_label;
-          update_block cfg handler_label handler_stack)
-        exceptional_successor
+          update_block cfg handler_label ~stack_offset ~traps:rest
+      end
     end
 
   let update_cfg : Cfg.t -> unit =
-   fun cfg -> update_block cfg cfg.entry_label []
+   fun cfg ->
```

Comment:
I think the main change is that the module
no longer relies on `can_raise` over basic
and terminator instructions. We should maybe
leave a comment to this effect. Indeed, we
will want to revisit the behavior as soon as
we decide it is not taboo to remove a dead
exception handler.

Reply:
I added a comment at the top the this module.